builtin/logical/aws/secret_access_keys_test.go

```diff
@@ -40,3 +45,123 @@ func TestNormalizeDisplayName_NormNotRequired(t *testing.T) {
 		}
 	}
 }
+
+func TestGenUsername(t *testing.T) {
+
+	testUsername, warning := genUsername("name1", "policy1", "iam_user", `{{ printf "vault-%s-%s-%s-%s" (.DisplayName) (.PolicyName) (unix_time) (random 20) | truncate 64 }}`)
+	if warning != "" {
+		t.Fatalf(
+			"expected no warning; got %s",
+			warning,
+		)
+	}
+
+	expectedUsernameRegex := `^vault-name1-policy1-[0-9]+-[a-zA-Z0-9]+`
+	require.Regexp(t, expectedUsernameRegex, testUsername)
+	// IAM usernames are capped at 64 characters
+	if len(testUsername) > 64 {
+		t.Fatalf(
+			"expected IAM username to be of length 64, got %d",
+			len(testUsername),
+		)
+	}
+
+	testUsername, warning = genUsername("name2", "policy2", "iam_user", `{{ printf "foo-%s-%s" (unix_time) (random 20) | truncate 64 }}`)
+	expectedUsernameRegex = `^foo-[0-9]+-[a-zA-Z0-9]+`
+	require.Regexp(t, expectedUsernameRegex, testUsername)
+
+	testUsername, warning = genUsername("name1", "policy1", "sts", defaultSTSTemplate)
+	if strings.Contains(testUsername, "name1") || strings.Contains(testUsername, "policy1") {
+		t.Fatalf(
+			"expected sts username to not contain display name or policy name; got %s",
+			testUsername,
+		)
+	}
+	// STS usernames are capped at 64 characters
+	if len(testUsername) > 32 {
+		t.Fatalf(
+			"expected sts username to be under 32 chars; got %s of length %d",
+			testUsername,
+			len(testUsername),
+		)
+	}
+}
+
+func TestReadConfig(t *testing.T) {
```

Comment:
nit: You could probably reduce some of the code here with "table driven tests". Some resources:
- short examples:
  - https://yourbasic.org/golang/table-driven-unit-test/
  - https://github.com/golang/go/wiki/TableDrivenTests
- long example: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

If you aren't going to use table driven tests, it might be good to at least separate each scenario into a separate unit test. For example, `TestReadConfig_DefaultTemplate` and `TestReadConfig_CustomTemplate`.

Of course these are just suggestions on style and organization. The code is functional as is.

Reply:
Thanks! That makes sense, I separated them for better readability :)